### PR TITLE
DisplayList tracks maximum render op depths

### PR DESCRIPTION
**Pull request by flar**
The DisplayList will now track the "render depth" of all rendering ops and provide the total content depth for DisplayList and save[Layer]/restore pairs to enable depth clipping in the back ends.

The depth accounting process is described in the class comment for `DlOpReceiver`

### CODE REVIEW

**Comment by flar** [comment]
I've updated the unit tests to double-check the max_depth values on the DisplayList in the existing unit tests. While I was in there I did some cleanup. There used to be 4 values on each "test snippet", as follows:

- display list op count
- display list byte count
- dl->skia picture->dl recapture op count
- dl->skia picture->dl recapture byte count

We haven't done "skia picture recapture" testing for a long time since we eliminated the need to convert Skia calls into DisplayList calls so the last 2 values have been obsolete for a while. I needed to add "depth" counts to all of the snippets, though, so while I was making that sweeping change I deleted the "sk recapture" numbers. The 3 numbers at the start of each snippet are now:

- display list op count
- display list byte count
- display list depth

While doing that, I also deleted all comments related to fudging the numbers to satisfy the Skia recapture logic in the tests.

**Comment by bdero** [comment]
I'm not sure there would be any utility in hooking this feature up to the current version of the DL dispatcher. We should hook this up to the new DL dispatcher as it's being formed, because
1. Aiks/EntityPass will need to be almost entirely rewritten anyhow,
2. we need to keep Aiks working without the DL until we're able to migrate them to use the DL.

File: display_list/display_list_unittests.cc [hunk 1]
```diff
@@ -3781,5 +3797,81 @@ TEST_F(DisplayListTest, SaveLayerBoundsClipDetectionSimpleClippedRect) {
...
+TEST_F(DisplayListTest, SaveContentDepthTest) {
+  DisplayListBuilder builder;
+  builder.DrawRect({10, 10, 20, 20}, DlPaint());  // depth 1
+
+  builder.Save();  // lasts through depth 6
+  {
+    builder.Translate(5, 5);
+    builder.DrawRect({10, 10, 20, 20}, DlPaint());  // depth 2
+
+    builder.SaveLayer(nullptr, nullptr);  // lasts through depth 4
```
Comment (bdero) [inline display_list/display_list_unittests.cc:3854]: Ah, this is something I should have specified. SaveLayers have a draw call call that also needs a unique depth. But they are drawn _after_ all of their children have been rendered, so the depth we use for them must be greater than all contained children drawn within the SaveLayers.

I recommend [replicating this test](https://github.com/flutter/engine/blob/main/impeller/aiks/aiks_unittests.cc#L2929-L2974) against the DL depth values to ensure that:
1. SaveLayers are reserved a depth value, and
2. clips interact correctly with SaveLayers.

Comment (flar) [reply display_list/display_list_unittests.cc:3854]: I noticed that while I was updating dl_dispatcher.cc...

The odd thing is that they only allocate the depth on the restore after recording the max_depth, right? The clips inside them will be rendered to the "restore copy - 1" max depth and then the restore copy happens above that depth so it is only clipped by the enclosing clips.

**Comment by flar** [comment]
> As an aside: I'm not sure there would be any utility in hooking this feature up to the current version of the DL dispatcher. We should hook this up to the new DL dispatcher as it's being formed, because
> 
> 1. Aiks/EntityPass will need to be almost entirely rewritten anyhow,
> 2. we need to keep Aiks working without the DL until we're able to migrate them to use the DL.

I prototyped it, but now I'm working through adjusting the new max_depth testing in the DL unit tests to account for saveLayer depths.

File: display_list/display_list_unittests.cc (hunk 1 above)
Comment (flar) [reply display_list/display_list_unittests.cc]: I think I've accounted for this in the latest commit.

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: display_list/display_list_unittests.cc [hunk 2]
```diff
@@ -3781,5 +3800,87 @@ TEST_F(DisplayListTest, SaveLayerBoundsClipDetectionSimpleClippedRect) {
...
+    }
+    builder.Restore();  // layer is restored with depth 7
+
+    builder.DrawRect({16, 16, 26, 26}, DlPaint());  // depth 8
+    builder.DrawRect({18, 18, 28, 28}, DlPaint());  // depth 9
+  }
+  builder.Restore();
+
+  builder.DrawRect({16, 16, 26, 26}, DlPaint());  // depth 10
+  builder.DrawRect({18, 18, 28, 28}, DlPaint());  // depth 11
```
Comment (bdero) [inline display_list/display_list_unittests.cc:3876]: Could you add a test that checks the depth of clip ops added to both the root and to the root and within a SaveLayer? That way we know they close with the correct depth (max depth of everything they affect).

Comment (flar) [reply display_list/display_list_unittests.cc:3876]: What is the "depth of clip ops"? They aren't accounted in DL. That's an Impeller thing...?

All DL does is count how many render calls are within a save/restore or an entire DL. Clips are invisible to that.

Comment (flar) [reply display_list/display_list_unittests.cc:3876]: As in the class comment in dl_op_receiver, the only places depth information is reported is the DL total_depth and the argument passed to save/saveLayer.

**Pull request merged**